src/v/config/constraint_types.cc

```diff
@@ -0,0 +1,703 @@
+          cluster_property);
+        return false;
+    }
+
+    if (enabled == constraint_enabled_t::no) {
+        // A constraint that is turned off (i.e., no) means that the topic
+        // property is automatically valid because the broker does not compare
+        // the topic and cluster values.
+        return true;
+    }
+
+    if (topic_val != cluster_val) {
+        vlog(
+          constraints_log.error,
+          "Constraints failure[does not match the cluster-level value {}]: "
+          "topic property {}.{}, value {}",
+          cluster_val,
+          topic(),
+          topic_property,
+          topic_val);
+        return false;
+    }
+
+    return true;
+}
+
+/**
+ * Returns true if the topic-level value satisfies the constraint.
+ * \param topic_val: the topic value
+ * \param constraint: the constraint to evaluate
+ * \param property_name: name of the cluster-level property
+ * \param cluster_opt: the value from the cluster-level property
+ */
+template<typename T>
+bool validate_value(
+  const std::optional<T>& topic_val,
+  const constraint_t& constraint,
+  const model::topic& topic,
+  const std::string_view& topic_property,
+  const std::string_view& cluster_property,
+  const std::optional<T> cluster_opt = std::nullopt) {
+    if (constraint.name == cluster_property) {
+        return ss::visit(
+          constraint.flags,
+          [&topic_val,
+           &topic,
+           &topic_property,
+           &cluster_property,
+           &cluster_opt](const constraint_enabled_t enabled) {
+              return matches_cluster_property_value(
+                topic_val,
+                cluster_opt,
+                enabled,
+                topic,
+                topic_property,
+                cluster_property);
+          },
+          [&topic_val, &topic, &topic_property](const auto range) {
+              return within_range(topic_val, range, topic, topic_property);
+          });
+    }
+    return false;
+}
+
+template<typename T>
+bool validate_value(
+  const tristate<T>& tri,
+  const constraint_t& constraint,
+  const model::topic& topic,
+  const std::string_view& topic_property,
+  const std::string_view& cluster_property,
+  const std::optional<T> cluster_opt = std::nullopt) {
+    if (tri.is_disabled()) {
+        return false;
+    }
+
+    return validate_value(
+      tri.get_optional(),
+      constraint,
+      topic,
+      topic_property,
+      cluster_property,
+      cluster_opt);
+}
+} // namespace
+
+bool topic_config_satisfies_constraint(
+  const cluster::topic_configuration& topic_cfg,
+  const constraint_t& constraint) {
+    auto partition_count_opt = std::make_optional(topic_cfg.partition_count);
+    auto replication_factor_opt = std::make_optional(
+      topic_cfg.replication_factor);
+    bool ret
+      = validate_value(
+          partition_count_opt,
+          constraint,
+          topic_cfg.tp_ns.tp,
+          kafka::topic_property_partition_count,
```

Comment:
The `config` library is rather foundational in Redpanda--everything includes it. I think we should catch this reverse dependency (along with the need for the kafka codegen to bring in the request types) before it gets baked in and becomes very hard to untangle later.